From 121e30e286344402af2a0996c1e6b3519c089852 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Tue, 11 Jun 2019 18:52:33 +0200
Subject: [PATCH 1/1] efi_loader: LoadImage w/o SourceBuffer and DevicePath

If both SourceBuffer and DevicePath are NULL, LoadImage() must return
EFI_INVALID_PARAMETER.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_boottime.c | 15 ++++-----------
 1 file changed, 4 insertions(+), 11 deletions(-)

diff --git a/lib/efi_loader/efi_boottime.c b/lib/efi_loader/efi_boottime.c
index fa01bbda70..7bb0fc2e52 100644
--- a/lib/efi_loader/efi_boottime.c
+++ b/lib/efi_loader/efi_boottime.c
@@ -1856,17 +1856,10 @@ efi_status_t EFIAPI efi_load_image(bool boot_policy,
 	EFI_ENTRY("%d, %p, %pD, %p, %zd, %p", boot_policy, parent_image,
 		  file_path, source_buffer, source_size, image_handle);
 
-	if (!image_handle || !efi_search_obj(parent_image)) {
-		ret = EFI_INVALID_PARAMETER;
-		goto error;
-	}
-
-	if (!source_buffer && !file_path) {
-		ret = EFI_NOT_FOUND;
-		goto error;
-	}
-	/* The parent image handle must refer to a loaded image */
-	if (!parent_image->type) {
+	if (!image_handle || (!source_buffer && !file_path) ||
+	    !efi_search_obj(parent_image) ||
+	    /* The parent image handle must refer to a loaded image */
+	    !parent_image->type) {
 		ret = EFI_INVALID_PARAMETER;
 		goto error;
 	}
-- 
2.20.1

